### DISCUSSION TITLE
trying to understand reconcile

### DISCUSSION THREAD

**Discussion by georgfaust** [solidjs/solid#1420] [Q&A]
I have a store like this (data is actually coming from an external observable, so I do not use the setStore path syntax but reconcile).

```
let initial_data = {
  a: { id: "a", text: "AAA" },
  b: { id: "b", text: "BBB" },
};
const [store, setStore] = createStore(initial_data);
```

and this `For`

```
<For each={Object.values(store).map((item) => ({ ...item }))}>
  {({ text }) => <div>{text}</div>}
</For>
```

if I mutate the data and then reconcile, `For` does not rerender.

```
initial_data.a.text = "AAA-RECONCILED";
setStore(reconcile(initial_data)); // :(
```

if I first clone the data it is rerendered:

```
let cloned_data = JSON.parse(JSON.stringify(initial_data));
cloned_data.a.text = "AAA-RECONCILED-CLONED";
setStore(reconcile(cloned_data)); // :)
```

Maybe I have to set `key` but what would it be in a nested store?
Or is it intended, that reconcile only checks if the ref of an object changed?
Can I force a diff somehow (the observable I'm using does mutate)

**Comment by atk** [comment]
Changing the initial data of a store instead of using the store setter will break things, because the store will use a comparison with the initial data object to detect changes. But then, `reconcile` is meant to turn immutable state into fine-grained reactive state. Imagine you have a JSON that you poll from a server that sometimes changes one or two of its properties. You can use `reconcile` to feed that into your store and not trigger updates on objects that got new references, but no changed data.

**Comment by georgfaust** [comment]
In this example I tried to rebuild whats happending with the library I'm using. It provides an observable I want to use in Solid, this is how I am using it.
```
externalStore.subscribe((update) => {
  if (update) {
    setMovables(reconcile(update));
  }
});
```
The first update would be like `initial_data` in the example. In all further updates the library just mutates the store. 

If reconcile expects JSON, that will obviously never be a problem, so I have to find another way. Just checking if I'm doing sth wrong.

**Reply by atk** [reply]
You have a superfluous `_` after update in line 3 of your code. Maybe that's the issue?

**Reply by georgfaust** [reply]
thanks for the pointer, that was just a copy-paste error.
